File: cpp/src/apu.cpp
/* apu.cpp — the T6W28. See apu.hpp for who writes to it and how I know.
 *
 * The register decode below mirrors core/apu.py line for line, because that model
 * is the ORACLE: the differential test drives both with the same byte stream and
 * demands the same state. If you change one, change the other, or the test fails
 * -- which is the point.
 */
#include "apu.hpp"

#include <algorithm>
#include <cstring>

namespace ngpc {

void Apu::reset() {
    *this = Apu();
}

int Apu::active_noise_period() const {
    if (noise.period_select < 3) return kNoisePeriods[noise.period_select];
    return noise.period_extra;
}

/* The two ports differ only in WHAT a non-volume command means:
 *   LEFT  -> tone periods (channels 0..2)
 *   RIGHT -> the noise control, and tone-3's period reused as the noise period
 * Volumes work on both, one stereo side each -- that IS the T6W28's stereo. */
void Apu::write(uint8_t data, bool left) {
    uint8_t& latch_reg = left ? latch_left : latch_right;
    const uint8_t latch = (data & 0x80) ? data : latch_reg;
    if (data & 0x80) latch_reg = data;

    const int index = (latch >> 5) & 3;

    if (latch & 0x10) {                                  /* volume */
        const int amp = kApuVolumes[data & 0x0F];
        if (index < 3) {
            if (left) square[index].vol_left = amp;
            else      square[index].vol_right = amp;
        } else {
            if (left) noise.vol_left = amp;
            else      noise.vol_right = amp;
        }
        return;
    }

    if (left) {
        if (index >= 3) return;                          /* no tone-3 on this chip */
        Square& sq = square[index];
        /* The 10-bit period arrives in two bytes and is stored ALREADY SHIFTED
         * LEFT BY FOUR -- i.e. multiplied by the chip's own /16 divider -- which is
         * why `period` is in chip clocks and why the mute threshold is 128 and not
         * 8. Do not "fix" the shift. */
        if (data & 0x80) sq.period = (sq.period & 0x3F00) | ((data << 4) & 0x00FF);
        else             sq.period = (sq.period & 0x00FF) | ((data << 8) & 0x3F00);
        return;
    }

    if (index == 2) {                                    /* the noise's extra period */
        if (data & 0x80)
            noise.period_extra = (noise.period_extra & 0x3F00) | ((data << 4) & 0x00FF);
        else
            noise.period_extra = (noise.period_extra & 0x00FF) | ((data << 8) & 0x3F00);
        return;
    }
    if (index == 3) {                                    /* the noise control */
        noise.period_select = data & 3;
        noise.tap     = (data & 0x04) ? kTapWhite : kTapDisabled;
        noise.shifter = 0x4000;                          /* every reconfig resets it */
    }
}

void Apu::write_left(uint8_t data)  { write(data, true); }
void Apu::write_right(uint8_t data) { write(data, false); }

/* One output sample. The oscillators are advanced a whole sample-period at a time
 * -- never clock by clock: at 3.072 MHz that would be 150 million steps a second
 * at our replay speed, and the emulator would stop being fast.
 *
 * This is a POINT sample, which the spec allows in as many words ("resampler naif")
 * and which aliases. Band-limited synthesis is a later pass, not a silent one. */
void Apu::emit_sample() {
    /* ⚠️ THE SAMPLE STEP IS FRACTIONAL, AND TRUNCATING IT DETUNES THE WHOLE CHIP.
     *
     * 3 072 000 / 44 100 = 69.66 chip clocks per output sample. This used to be an
     * integer divide -- 69 -- so the oscillators advanced 0.95 % too slowly, every
     * note came out 0.95 % sharp, and the audio clock ran at 100.9 % of real time
     * (measured). A drift that small is inaudible as pitch and lethal as timing: it
     * overruns the host's audio buffer a little more every second.
     *
     * Fixed point, 16.16. The remainder is carried, so the error does not accumulate. */
    static constexpr uint32_t kStepFP =
        uint32_t((uint64_t(kApuClockHz) << 16) / kAudioSampleHz);   /* 69.66 in 16.16 */
    step_fp += kStepFP;
    const uint32_t step = step_fp >> 16;
    step_fp &= 0xFFFF;

    int left = 0, right = 0;

    for (Square& sq : square) {
        if (sq.period <= kMinAudiblePeriod || (!sq.vol_left && !sq.vol_right)) continue;
        sq.counter += int(step);
        const int toggles = sq.counter / sq.period;
        sq.counter %= sq.period;
        sq.phase ^= (toggles & 1);
        const int sign = sq.phase ? 1 : -1;
        left  += sign * sq.vol_left;
        right += sign * sq.vol_right;
    }

    if (noise.vol_left || noise.vol_right) {
        const int period = std::max(1, 2 * active_noise_period());
        noise.counter += int(step);
        int steps = noise.counter / period;
        noise.counter %= period;
        /* Bounded: a tiny extra-period could otherwise ask for thousands of LFSR
         * steps per sample. 64 is far more than any real period needs. */
        steps = std::min(steps, 64);
        for (int i = 0; i < steps; ++i) {
            noise.shifter = (((noise.shifter << 14) ^ (noise.shifter << noise.tap)) & 0x4000)
                          | (noise.shifter >> 1);
        }
        const int sign = (noise.shifter & 1) ? -1 : 1;
        left  += sign * noise.vol_left;
        right += sign * noise.vol_right;
    }

    /* ⚡ THE SAMPLED VOICE, SUMMED IN. The DAC bypasses the sound chip entirely -- it is
     * the main CPU driving the speaker directly, one byte at a time -- so it is added to
     * the chip's output, not routed through it. Held between writes: the converter keeps
     * driving the last code it was handed. 0x80 is silence, so a game that never touches
     * the DAC contributes exactly nothing here. */
    left  += (int(dac_left)  - kDacSilence) * kDacGain;
    right += (int(dac_right) - kDacSilence) * kDacGain;

    /* Four channels at 64 each = +-256 worst case; scale to a comfortable
     * headroom rather than clipping the one frame where they all align. */
    const int16_t l = int16_t(std::clamp(left  * 64, -32768, 32767));
    const int16_t r = int16_t(std::clamp(right * 64, -32768, 32767));

    if (produced - drained >= kRingFrames) {   /* the host fell behind */
        ++drained;                             /* drop the OLDEST, and COUNT it */
        ++dropped;
    }
    const uint32_t slot = uint32_t(produced % kRingFrames) * 2;
    ring[slot]     = l;
    ring[slot + 1] = r;
    ++produced;
}

void Apu::tick(uint32_t main_cycles) {
    if (main_cycles == 0) return;

    /* main clock -> chip clock (half), carrying the remainder so the audio clock
     * cannot drift away from the video clock over a long session. */
    main_residue += main_cycles;
    const uint32_t chip = main_residue / 2;
    main_residue %= 2;
    if (chip == 0) return;

    /* How many output samples does `chip` chip-clocks buy? Exactly
     * `chip * 44100 / 3072000`, with the remainder carried -- NOT `chip / 69`, which
     * is what this did and which produced 44 522 Hz worth of samples per second while
     * claiming 44 100. */
    chip_residue += chip * kAudioSampleHz;
    uint32_t samples = chip_residue / kApuClockHz;
    chip_residue %= kApuClockHz;

    /* A single `ldir` can bill twenty thousand cycles; a HALT idles a whole frame.
     * Cap the burst so one pathological instruction cannot ask for a million
     * samples in one call -- the ring only holds 16 384 anyway. */
    samples = std::min(samples, Apu::kRingFrames);
    for (uint32_t i = 0; i < samples; ++i) emit_sample();
}

uint32_t Apu::drain(int16_t* out, uint32_t n) {
    const uint64_t available = produced - drained;
    const uint32_t want = uint32_t(std::min<uint64_t>(available, n));
    for (uint32_t i = 0; i < want; ++i) {
        const uint32_t slot = uint32_t((drained + i) % kRingFrames) * 2;
        out[i * 2]     = ring[slot];
        out[i * 2 + 1] = ring[slot + 1];
    }
    drained += want;
    return want;
}

}  // namespace ngpc

File: core/apu.py
"""T6W28 sound chip oracle (clean-room) for NgpCraft Emulator.

Immutable, deterministic model of the NGPC T6W28 APU: three square (tone)
oscillators plus one noise oscillator, with *independent* left/right volume
latches (the stereo feature that distinguishes the T6W28 from a plain
SN76489).

This module is the reference oracle described in ``specs/APU_T6W28.md``.
It is written from scratch from that spec (facts + algorithm), not ported
from any LGPL source. Same design contract as ``core/cpu.py``: frozen
dataclasses and pure ``write_*`` / ``run_until`` functions that return a new
state instead of mutating in place.

Signal reconstruction here is deliberately naive (per-cycle bipolar
amplitude), enough for an opcode/register oracle. Band-limited synthesis is
a job for the future real-time C++ core, not for this reference model.
"""

from __future__ import annotations

from dataclasses import dataclass, replace

# Logarithmic attenuation table, index 0 = full, 15 = silent.
# volumes[i] = round(64 * 1.26 ** (15 - i) / 1.26 ** 15)
VOLUMES: tuple[int, ...] = (64, 50, 39, 31, 24, 19, 15, 12, 9, 7, 5, 4, 3, 2, 1, 0)

# Fixed noise divisor table selected by (data & 3) in {0, 1, 2}.
NOISE_PERIODS: tuple[int, ...] = (0x100, 0x200, 0x400)

# Tap position of the LFSR: 13 = white noise, 16 = tap disabled (periodic).
TAP_WHITE = 13
TAP_DISABLED = 16


@dataclass(frozen=True)
class Oscillator:
    """One of the four T6W28 oscillators (3 squares + 1 noise).

    ``volume_left`` / ``volume_right`` are post-table amplitudes (0..64).
    ``phase`` is the square-wave phase bit (unused by the noise osc).
    """

    volume_left: int = 0
    volume_right: int = 0
    period: int = 0
    phase: int = 0


@dataclass(frozen=True)
class NoiseState:
    """LFSR noise generator state."""

    shifter: int = 0x4000
    tap: int = TAP_WHITE
    period_select: int = 0  # 0..2 -> NOISE_PERIODS, 3 -> period_extra
    period_extra: int = 0
    volume_left: int = 0
    volume_right: int = 0


@dataclass(frozen=True)
class ApuState:
    """Full immutable T6W28 state.

    ``squares`` holds oscillators 0..2 (tone), ``noise`` is oscillator 3.
    ``latch_left`` / ``latch_right`` remember the last command byte written
    to each stereo port so a following data byte reuses its channel index.
    """

    squares: tuple[Oscillator, Oscillator, Oscillator] = (
        Oscillator(),
        Oscillator(),
        Oscillator(),
    )
    noise: NoiseState = NoiseState()
    latch_left: int = 0
    latch_right: int = 0


def reset_state() -> ApuState:
    """Return the cold-start APU image."""

    return ApuState()


def _active_noise_period(noise: NoiseState) -> int:
    if noise.period_select < 3:
        return NOISE_PERIODS[noise.period_select]
    return noise.period_extra


def _set_square_volume_left(state: ApuState, index: int, amp: int) -> ApuState:
    sq = replace(state.squares[index], volume_left=amp)
    squares = tuple(sq if i == index else s for i, s in enumerate(state.squares))
    return replace(state, squares=squares)  # type: ignore[arg-type]


def _set_square_volume_right(state: ApuState, index: int, amp: int) -> ApuState:
    sq = replace(state.squares[index], volume_right=amp)
    squares = tuple(sq if i == index else s for i, s in enumerate(state.squares))
    return replace(state, squares=squares)  # type: ignore[arg-type]


def _set_volume(state: ApuState, index: int, amp: int, *, left: bool) -> ApuState:
    """Set post-table volume for channel ``index`` on one stereo side."""

    if index < 3:
        if left:
            return _set_square_volume_left(state, index, amp)
        return _set_square_volume_right(state, index, amp)
    # noise oscillator (index 3)
    if left:
        return replace(state, noise=replace(state.noise, volume_left=amp))
    return replace(state, noise=replace(state.noise, volume_right=amp))


def _write_square_period(state: ApuState, index: int, data: int) -> ApuState:
    """Two-byte period build for a tone channel (14-bit, 0..0x3FFF)."""

    sq = state.squares[index]
    if data & 0x80:  # latch byte -> low nibble
        period = (sq.period & 0x3F00) | ((data << 4) & 0x00FF)
    else:  # data byte -> high 6 bits
        period = (sq.period & 0x00FF) | ((data << 8) & 0x3F00)
    squares = tuple(
        replace(sq, period=period) if i == index else s
        for i, s in enumerate(state.squares)
    )
    return replace(state, squares=squares)  # type: ignore[arg-type]


def write_left(state: ApuState, data: int) -> ApuState:
    """Handle a write to the LEFT sound port (NGPC I/O 0xA1).

    Sets tone periods and per-channel *left* volumes.
    """

    data &= 0xFF
    latch = data if (data & 0x80) else state.latch_left
    if data & 0x80:
        state = replace(state, latch_left=data)
    index = (latch >> 5) & 3
    if latch & 0x10:
        return _set_volume(state, index, VOLUMES[data & 0x0F], left=True)
    if index < 3:
        return _write_square_period(state, index, data)
    return state


def write_right(state: ApuState, data: int) -> ApuState:
    """Handle a write to the RIGHT sound port (NGPC I/O 0xA0).

    Sets per-channel *right* volumes plus the noise control / extra period.
    """

    data &= 0xFF
    latch = data if (data & 0x80) else state.latch_right
    if data & 0x80:
        state = replace(state, latch_right=data)
    index = (latch >> 5) & 3
    if latch & 0x10:
        return _set_volume(state, index, VOLUMES[data & 0x0F], left=False)
    if index == 2:
        # extra noise period, shares tone-3's two-byte build.
        noise = state.noise
        if data & 0x80:
            extra = (noise.period_extra & 0x3F00) | ((data << 4) & 0x00FF)
        else:
            extra = (noise.period_extra & 0x00FF) | ((data << 8) & 0x3F00)
        return replace(state, noise=replace(noise, period_extra=extra))
    if index == 3:
        select = data & 3
        tap = TAP_WHITE if (data & 0x04) else TAP_DISABLED
        return replace(
            state,
            noise=replace(state.noise, period_select=select, tap=tap, shifter=0x4000),
        )
    return state


def _lfsr_step(shifter: int, tap: int) -> tuple[int, bool]:
    """Advance the 15-bit LFSR one step; return (new_shifter, output_changed)."""

    changed = bool((shifter + 1) & 2)
    shifter = (((shifter << 14) ^ (shifter << tap)) & 0x4000) | (shifter >> 1)
    return shifter, changed


def run_until(state: ApuState, cycles: int) -> tuple[ApuState, list[tuple[int, int]]]:
    """Run all oscillators for ``cycles`` chip clocks.

    Returns the advanced state plus a list of ``(left, right)`` bipolar
    amplitude samples, one per chip clock. This is the naive oracle output;
    it is not band-limited.
    """

    if cycles <= 0:
        return state, []

    squares = list(state.squares)
    noise = state.noise
    noise_period = max(1, 2 * _active_noise_period(noise))
    samples: list[tuple[int, int]] = []

    for t in range(cycles):
        left = 0
        right = 0

        # Tone channels: toggle phase every `period` clocks; muted when
        # both volumes are 0 or the period is a near-ultrasonic <= 128.
        for i, sq in enumerate(squares):
            if sq.period > 128 and (sq.volume_left or sq.volume_right):
                if sq.period and (t % sq.period) == 0 and t > 0:
                    sq = replace(sq, phase=sq.phase ^ 1)
                    squares[i] = sq
                sign = 1 if sq.phase else -1
                left += sign * sq.volume_left
                right += sign * sq.volume_right

        # Noise channel.
        if noise.volume_left or noise.volume_right:
            sign = -1 if (noise.shifter & 1) else 1
            left += sign * noise.volume_left
            right += sign * noise.volume_right
            if t > 0 and (t % noise_period) == 0:
                shifter, _ = _lfsr_step(noise.shifter, noise.tap)
                noise = replace(noise, shifter=shifter)

        samples.append((left, right))

    new_state = replace(state, squares=tuple(squares), noise=noise)  # type: ignore[arg-type]
    return new_state, samples
